File: 프로그래머스/lv2/김현우/괄호 회전하기.py
from collections import deque

# 스택을 이용한 올바른 괄호 문자열 판별
def is_correct_string(w):
    s = []

    for i in range(len(w)):
        if w[i] == '{' or w[i] == '[' or w[i] == '(':
            s.append(w[i])
        else:
            if not s:
                return False
            top = s.pop()
            if w[i] == ']' and top != '[':
                return False
            if w[i] == '}' and top != '{':
                return False
            if w[i] == ')' and top != '(':
                return False

    if len(s) > 0:
        return False
    return True

# 큐를 사용하여 모든 회전 경우의 수에 대하여 올바른 괄호 문자열 개수 카운트
def solution(s):
    answer = 0
    w = deque(s)

    for i in range(len(s)):
        if is_correct_string(w):
            answer += 1
        w.append(w.popleft())

    return answer

File: 프로그래머스/lv2/김원래/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(string s) {
    int answer = 0;
    
    int n = s.size();
    
    for(int x = 0 ; x < s.size() ; ++x)
    {
        stack<char> st;
        
        int cnt = n;
        
        int pos = x;
        
        bool flag = true;
        
        
        while(cnt--)
        {
            if(s[pos] == '[' || s[pos] == '(' || s[pos] == '{') {
                st.push(s[pos]);
            }
            else {
                if(st.empty()) {
                    flag = false;
                    break;
                }
                
                char top = st.top();
                
                st.pop();
                
                if(top == '[') {
                    if(s[pos] != ']') {
                        flag = false;
                        break;
                    }
                }
                else if(top == '(') {
                    if(s[pos] != ')') {
                        flag = false;
                        break;
                    }
                }
                else {
                    if(s[pos] != '}') {
                        flag = false;
                        break;
                    }
                }
            }
            pos = (pos + 1) % n;
        }
        
        if(!st.empty()) continue;
        
        if(flag == true) ++answer;
    }
    
    return answer;
}

/*


*/